internal/controller/dashboard_controller.go

```diff
@@ -0,0 +1,345 @@
+// reconcileCreate ensures the Grafana dashboard described in configmap is created in Grafana.
+// This function is also responsible for:
+// - Adding the finalizer to the configmap
+func (r DashboardReconciler) reconcileCreate(ctx context.Context, dashboard *v1.ConfigMap) (ctrl.Result, error) { // nolint:unparam
+	logger := log.FromContext(ctx)
+
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	if !controllerutil.ContainsFinalizer(dashboard, DashboardFinalizer) {
+		// We use a patch rather than an update to avoid conflicts when multiple controllers are adding their finalizer to the grafana dashboard
+		// We use the patch from sigs.k8s.io/cluster-api/util/patch to handle the patching without conflicts
+		logger.Info("adding finalizer", "finalizer", DashboardFinalizer)
+		patchHelper, err := patch.NewHelper(dashboard, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+		controllerutil.AddFinalizer(dashboard, DashboardFinalizer)
+		if err := patchHelper.Patch(ctx, dashboard); err != nil {
+			logger.Error(err, "failed to add finalizer", "finalizer", DashboardFinalizer)
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+		logger.Info("added finalizer", "finalizer", DashboardFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Configure the dashboard in Grafana
+	if err := r.configureDashboard(ctx, dashboard); err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func getDashboardUID(dashboard map[string]interface{}) (string, error) {
+	UID, ok := dashboard["uid"].(string)
+	if !ok {
+		return "", errors.New("dashboard UID not found in configmap")
+	}
+	return UID, nil
+}
+
+func getDashboardCMOrg(dashboard *v1.ConfigMap) (string, error) {
+	// Try to look for an annotation first
+	annotations := dashboard.GetAnnotations()
+	if annotations != nil && annotations[organization.OrganizationLabel] != "" {
+		return annotations[organization.OrganizationLabel], nil
+	}
+
+	// Then look for a label
+	labels := dashboard.GetLabels()
+	if labels != nil && labels[organization.OrganizationLabel] != "" {
+		return labels[organization.OrganizationLabel], nil
+	}
+
+	// Return an error if no label was found
+	return "", errors.New("No organization label found in configmap")
+}
+
+func (r DashboardReconciler) configureDashboard(ctx context.Context, dashboardCM *v1.ConfigMap) error {
+	logger := log.FromContext(ctx)
+
+	dashboardOrg, err := getDashboardCMOrg(dashboardCM)
+	if err != nil {
+		logger.Info("Skipping dashboard, no organization found")
+		return nil
+	}
+
+	// We always switch back to the shared org
+	defer func() {
+		if _, err = r.GrafanaAPI.SignedInUser.UserSetUsingOrg(grafana.SharedOrg.ID); err != nil {
+			logger.Error(err, "failed to change current org for signed in user")
+		}
+	}()
+
+	// Switch context to the dashboards-defined org
+	organization, err := grafana.FindOrgByName(r.GrafanaAPI, dashboardOrg)
```

Comment:
Here, I'm not sure if the name should be the GrafanaOrganization CR `metadata.name` as this field cannot change (at least not without recreating the whole org) or GrafanaOrganization CR `spec.displayName` as this value can be updated by users.

Reply:
I asked myself the same question.

The displayName (as used currently) seems more flexible, as it's not tied to the way we create orgs.
It allows to load dashboards in orgs that don't have a CR, ie orgs that were created via other ways like clickops.
In the future, we can change the way we create the orgs without breaking dashboards loading.

Reply:
sure